agent/agent.go

```diff
@@ -156,18 +174,13 @@ type agent struct {
 	// filter out events that have already been recorded by this agent.
 	lastSeen *ttlmap.TTLMap
 
-	// Config is the agent configuration.
-	Config
-
-	// ClusterMembership provides access to cluster membership service.
-	ClusterMembership membership.ClusterMembership
-
-	// ClusterTimeline keeps track of all timeline events in the cluster. This
-	// timeline is only used by members that have the role 'master'.
-	ClusterTimeline history.Timeline
-
-	// LocalTimeline keeps track of local timeline events.
-	LocalTimeline history.Timeline
+	// cancel is used to cancel the internal processes
+	// running as part of g
+	cancel context.CancelFunc
+	// ctx manages the lifetime of the agent's internal processes
+	ctx context.Context
```

Comment:
nit: I know it makes it easier, but I thought the idiomatic pattern for context is that it should not be placed within a struct?

Reply:
It is not always possible to do that - in this case, since I went with `errgroup`, and additionally wanted one central group for the agent. Now I'm not able to access the original context the group was created with since the functions I pass to `errgroup.Go` do not accept any context and the errgroup API does not provide access to the underlying context - which is unfortunate.
`errgroup` seems to be the go-to package w.r.t handling groups of goroutines so I consider it an upgrade.
[Here](https://github.com/golang/go/issues/34510) is a relevant discussion about this shortcoming which is still unresolved.